### PR TITLE
Accessible hashes

### PR DESCRIPTION
**Pull request by aars**
This PR makes accessible:
`@attributes_hash`, `@links_hash` and `@relationships_hash` from `JSON::API::Resource`
`@data_hash` and `@links_hash` from `JSON::API::Relationship`

Motivation:
Hash API is way more generic and can be used directly by `Sequel` or `ActiveRecord` for example. Getting these hashes from the `JSON::API::Document` is much nicer than keeping both `JSON::API::Document` and whatever hash it originally parsed in your application. You can now pass around `JSON::API::Document` objects within your application without having to access `@the_original_hash_or_json_string` as well.

Tests added. Let me know if there is something to be improved or clarified.

### CODE REVIEW

**Comment by beauby** [comment]
Hi @aars, thanks for the PR!
A few remarks: 
+ I'd rather have a `to_hash` method on the `Attributes`/`Links`/`Relationships` classes (this is just a naming concern).
+ You might be interested in a proof-of-concept gem I made to tailor `jsonapi_parser` to Rails needs: [jsonapi_parser_rails](https://github.com/beauby/jsonapi_parser_rails) (which currently just enriches the `JSON::API::Resource` class with a `to_active_record_hash` method that takes care of mapping singular/plural relationships to `#{key}_id`/`#{key}_ids`.
Regardless, I think you're right, it makes sense keeping the initial hash around for each JSON API object to allow maximum flexibility.

TL;DR: Looks good for merging once the methods have been moved to the collection objects as `to_hash`.

**Comment by aars** [comment]
I've updated the PR. Have a look.

As you might have guessed, I'm not (yet) fluent in Ruby and it's idioms/patterns. I tried to find an elegant solution, let me know what you think and how I can improve it.

Thanks for linking the `to_active_record_hash`, that gives me some clues on how I could handle the same thing. But I'll figure that out when I need it. Currently I have an Ember Data client and I'm implementing it's requests and requirements as I go. Haven't looked into how it wants to update relationships (or in how far this is manual labor on either client- or server side)

Cheers

File: lib/json/api/to_hash.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module JSON
+  module API
+    module ToHash
+      def to_hash
```
Comment (beauby) [inline lib/json/api/to_hash.rb]: Mixing-in modules is definitely a cool feature of ruby, but in this specific case I don't think it is the right choice (reduces code clarity, might lead to bugs if the hash's name gets changed for some reason).

File: spec/parser_spec.rb [hunk 2]
```diff
@@ -2,37 +2,41 @@
```
Comment (beauby) [inline spec/parser_spec.rb]: :+1: for providing relevant tests!

File: lib/json/api/resource_identifier_array.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+module JSON
+  module API
+    class ResourceIdentifierArray < Array
```
Comment (beauby) [inline lib/json/api/resource_identifier_array.rb]: Nice :+1: Would like to rename it to `ResourceIdentifierCollection` though.

File: lib/json/api/resource_array.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+module JSON
+  module API
+    class ResourceArray < Array
```
Comment (beauby) [inline lib/json/api/resource_array.rb]: As well, would like to rename it to `ResourceCollection`.

**Comment by aars** [comment]
Renamed `*Array` to `*Collection`

About the mixin: If the hash name changes the `to_hash` would always have to change, no matter the implementation, right?

Would you prefer a simple `def to_hash` in the appropriate classes? 
I know the `ToHash` module doesn't add much (now), but this does force the separation of "a class having the `to_hash` feature" and "this is what the class has available to hash". 

In the current implementation we're exposing the original unfiltered, unmodified input. That might not always be the best idea when the `JSON::API` classes start to filter or represent the original input differently. You would probably want the `to_hash` to output the same (re-)representation.
The classes itself would then be concerned with providing the hash (be it new or old), and `JSON::API::ToHash` would provide an interface for outputting the provided hash (`@exposed_hash`, or maybe a method?). 
It would be the interface that your `jsonapi_parser_rails` would use for example.

Much cleaner than having `def to_hash` floating around every other class IMHO.

Or, it could very well be a case of premature overengineering :)

What would you like me to do?

**Comment by aars** [comment]
Merged master and resolved conflicts.

**Comment by beauby** [comment]
For now it does seem like over-engineering to me. Plus, we might want to support building JSON API objects in other ways than by parsing a payload, in which case the `to_hash` method would have to be customized for each object. 
So let's have it as simple methods for now, and if it makes sense to factor it out into a mixin down the line, we'll do it then.

**Comment by aars** [comment]
Coming up!

**Comment by aars** [comment]
Done!

**Comment by aars** [comment]
Current implementation ok?

**Comment by aars** [comment]
@beauby Ping!

**Comment by beauby** [comment]
Hey, sorry for the delay, I have been unexpectedly busy lately. Would you mind giving me commit rights on your fork so that I can make some final cosmetic changes and get this merged?

**Comment by aars** [comment]
@beauby I've added you as a collaborator. That should be enough, right?

(also sorry for the delay ;) I've been sick.)

**Comment by beauby** [comment]
Rebased and squashed, will add a few minor changes before merging (probably later tonight or tomorrow). Thanks @aars!

**Comment by aars** [comment]
Awesomesauce!

**Comment by beauby** [comment]
I haven't forgotten about this, will take care of it later tonight!

**Comment by beauby** [comment]
Took longer than expected, but we're finally there! Thanks for your work @aars.
